### ISSUE TITLE
json: cannot unmarshal object into Go value of type []string

### ISSUE THREAD

**Issue by ambroisemaupate** [stage1/docker-php#122]
Hi, 

I’ve got a `json: cannot unmarshal object into Go value of type []string` error only when I’m trying to `run` a new container with *Docker-PHP*.

Here is my Docker version:
```
Docker version 1.7.1, build 786b29d
```

I’m running my application in a Docker container which use the docker socket as volume.
```
    "Volumes": {
        "/data": "/var/www/littleship",
        "/var/run/docker.sock": "/run/docker.sock"
    },
    "VolumesRW": {
        "/data": true,
        "/var/run/docker.sock": true
    },  
```

And here is my Symfony action to create a container (https://github.com/ambroisemaupate/littleship/blob/master/src/AM/Bundle/DockerBundle/Controller/ContainerController.php#L144):

```
    public function addAction(Request $request)
    {
        if (!$this->isGranted('ROLE_SUPER_ADMIN')) {
            throw $this->createAccessDeniedException();
        }

        try {
            $docker = $this->get('docker');
            $cManager = $docker->getContainerManager();
            $iManager = $docker->getImageManager();

            $form = $this->createForm(new ContainerType($iManager, $cManager));

            $form->handleRequest($request);

            if ($form->isValid()) {
                $data = $form->getData();
                $newCont = new Container([
                    'HostConfig' => [
                        'Links' => $data['links'],
                        'VolumesFrom' => $data['volumes_from'],
                        'PublishAllPorts' => (boolean) $data['publish_ports'],
                        'RestartPolicy' => [
                            'Name' => $data['restart_policy'],
                            'MaximumRetryCount' => 0
                        ]
                    ]
                ]);
                $newCont->setImage($data['image']);
                $newCont->setExposedPorts($this->getExposedPorts($data['ports']));
                $newCont->setName($data['name']);
                $newCont->setEnv($data['env']);
                $cManager->run($newCont, null, [], true);

                return $this->redirect($this->generateUrl('am_docker_container_list'));
            }

            $assignation['form'] = $form->createView();
        } catch (RequestException $e) {
            $assignation['error'] = $e->getMessage();
        }

        return $this->render('AMDockerBundle:Container:add.html.twig', $assignation);
    }
```

This action perfectly worked before Docker 1.7.1 upgrade, and other actions still work (stop, restart  and remove containers). Do you have any idea ? Do you need more informations ?

Thanks!
Sincerely,

Ambroise

**Comment by akkonrad** [comment]
I had the same issue when calling container controller with array arguments, eg: 
```php
$container = new Container([
        'HostConfig' => [
          'Links' => ['db' => 'db'],
        ]
      ]);

// I think correct form should be:

    $container = new Container([
        'HostConfig' => [
          'Links' => ['db:db'],
        ]
      ]);

```

Now when I inspect container in console I get
```
docker inspect -f "{{ .HostConfig.Links }}" web
[/db:/qw/db]
```

So I guess that you messed something in array that is passed to the constructor

**Comment by boonkerz** [comment]
I have the same issue:
`
$container = new Container([
 'Image' => 'tutum/mysql',
 'Volumes' => [ '/var/lib/mysql'],
 'Env' => ['MYSQL_PASS=pass', 'MYSQL_USER=user']
]);
`

**Comment by Boran** [comment]
This has just happened to me when creating containers, after upgrade docker 1.6 to 1.7.1.
Quick analysis:
Something has changed in how docker handles volumes. If $manager->create($container) is called without volumes  no error happens.
In my instance the error happens on Http/Adapter/DockerAdapter line 77, whiich is the catch after
```
            RequestEvents::emitBefore($transaction);
            if (!$transaction->getResponse()) {
                $this->createResponse($transaction);
                RequestEvents::emitComplete($transaction);
            }
```
I dont see any volume formatting changes on http://docs.docker.com/reference/api/docker_remote_api/#what-s-new

**Comment by ambroisemaupate** [comment]
I think I’ve understood where the problem comes from. 

https://github.com/ambroisemaupate/littleship/commit/decefe1d29c372dd16919c9eba29cb41d49e4d0a#diff-12d54e6d79b2203970e6c68aa78be01bR139

With new Docker API, you cannot set **empty** arrays to your container configuration.
I’ve checked if `env`, `links`, `exposedPorts` and `volumes_from` data were not empty before setting them in `Container` object and running it. It solved my issue.

**Comment by hacfi** [comment]
@ambroisemaupate Thanks for the update..good to know!!

**Comment by Boran** [comment]
My issue was slight different from @ambroisemaupate  it seems.

One has to send empty arrays sometimes. To create a countainer that mounts, for example,  /opt/server/foo as /data in the container.

@boonkerz indicates how to send an empty array above..

When creating a container one does a create() and then start().
On create() send a config array, with a Volume array of index '/data' containing an empty array.
['Image'=> $contimage, 'Hostname' => $fqdn,  'Env'  => $dockerenv,  'Volumes'  => [ '/data' => array() ],    ];

Then for start(), send an array  containing a Binds array with one entry "/opt/server/foo:/datal:rw"

At least that works for me...

**Comment by thaJeztah** [comment]
if this is a breaking change in the Docker API, please open an issue in the Docker issue tracker.

It's possible that this has already been fixed in Docker 1.8-rc2, so it could be worth testing against that version. Details can be found here: https://github.com/docker/docker/pull/15091#issuecomment-127461032

**Comment by Boran** [comment]
I don't consider that it changes the API, it's just more precise in what is expects.  A string like "/data =>{}" probably should never have worked. We can close this issue at this stage?

**Comment by joelwurtz** [comment]
There is still a bug in the library we don't differentiate an empty array and an empty object as our library will do this by default : 

Json::encode(array()) => "{}"

but sometimes we need to have

Json::encode(array()) => "[]"

Thanks for the report, i will try to have a look at it (certainly moving from array to stdClass for json object and use array as json array and not object)

**Comment by joelwurtz** [comment]
closed in #147
